service/worker/deployment/deployment_util.go

```diff
@@ -111,3 +115,34 @@ func GenerateDeploymentWorkflowID(seriesName string, buildID string) string {
 
 	return DeploymentWorkflowIDPrefix + DeploymentWorkflowIDDelimeter + escapedSeriesName + DeploymentWorkflowIDDelimeter + escapedBuildId
 }
+
+func GenerateDeploymentWorkflowIDForPatternMatching(seriesName string) string {
+	// escaping the reserved workflow delimiter (|) from the inputs, if present
+	escapedSeriesName := escapeChar(seriesName)
+
+	return DeploymentWorkflowIDPrefix + DeploymentWorkflowIDDelimeter + escapedSeriesName + DeploymentWorkflowIDDelimeter
+}
+
+// BuildQueryWithSeriesFilter is a helper which builds a query for pattern matching based on the
+// provided seriesName
+func BuildQueryWithSeriesFilter(seriesName string) string {
+	wildCardedWorkflowID := GenerateDeploymentWorkflowIDForPatternMatching(seriesName)
+	query := fmt.Sprintf("%s AND %s STARTS_WITH '%s'", DeploymentVisibilityBaseListQuery, searchattribute.WorkflowID, wildCardedWorkflowID)
+	return query
+}
+
+// ParseDeploymentWorkflowID is a helper which parses out the deployment workflow ID to
+// extract the seriesName and buildID
+func ParseDeploymentWorkflowID(workflowID string) (string, string) {
```

Comment:
do not parse the workflow id ever. this is just wrong if the deployment name or build id contains the delimiter. put the fields in the memo if you need to.

Reply:
but are we not escaping the delimiter while building the ID?

Reply:
evaluate `ParseDeploymentWorkflowID(GenerateDeploymentWorkflowID("a|b", "c|d"))`

Reply:
now that you mentioned this - I noticed something very intriguing - 

I might be wrong here, but does ES have an issue when using escape characters in `STARTS_WITH` / `BETWEEN`?

I tried using the example you had cited by having series and deployment name equating to "a|b" and "c|d". I then tried pattern matching `WorkflowID` STARTS_WITH `DeploymentWorkflowIDPrefix + DeploymentWorkflowIDDelimeter + escapedSeriesName + DeploymentWorkflowIDDelimeter` which resulted in visibility not finding any executions!


I went down a rabit hole of verifying if both the workflowID and the generated wildcard match wrt to the escape characters and they do. Hence, I wonder if it's the presence of an escape character in our query we need to be worried about here...(I am still digging to find the root cause of this issue)

Reply:
@ShahabT and I think this *may* be a visibility concern since even after escaping the backslash while generating a query, list deployments seems to fail. For now, I propose replacing adding escapes by appending a delimiter as and when we see one in our inputs